### PR TITLE
[libc] Scanf shouldn't match just "0x" for hex int

### PR DESCRIPTION
**Pull request by michaelrj-google**
Scanf parsing reads the longest possibly valid prefix for a given
conversion. Then, it performs the conversion on that string. In the case
of "0xZ" with a hex conversion (either "%x" or "%i") the longest
possibly valid prefix is "0x", which makes it the "input item" (per the
standard). The sequence "0x" is not a "matching sequence" for a hex
conversion, meaning it results in a matching failure, and parsing ends.
This is because to know that there's no valid digit after "0x" it reads
the 'Z', but it can only put back one character (the 'Z') leaving it
with consuming an invalid sequence.

(inspired by a thread on the libc-coord mailing list: 
https://www.openwall.com/lists/libc-coord/2024/10/15/1, see 7.32.6.2 in
the standard for more details.)

### CODE REVIEW

File: libc/src/stdio/scanf_core/int_converter.cpp [hunk 1]
```diff
@@ -124,13 +124,25 @@ int convert_int(Reader *reader, const FormatSection &to_conv) {
...
+        // the execution of the directive fails: this condition is a matching
+        // failure"
         base = 16;
         if (max_width > 1) {
           --max_width;
           cur_char = reader->getc();
         } else {
-          write_int_with_length(0, to_conv);
-          return READ_OK;
+          // write_int_with_length(0, to_conv);
```
Comment (nickdesaulniers) [inline libc/src/stdio/scanf_core/int_converter.cpp:144]: Did you mean to remove this, rather than just commenting it out?

**Review by nickdesaulniers — COMMENTED** [review]
Consider linking to https://www.openwall.com/lists/libc-coord/2024/10/15/1 in the commit message.

File: libc/src/stdio/scanf_core/int_converter.cpp [hunk 2]
```diff
@@ -124,13 +124,24 @@ int convert_int(Reader *reader, const FormatSection &to_conv) {
...
+        // the execution of the directive fails: this condition is a matching
+        // failure"
         base = 16;
         if (max_width > 1) {
           --max_width;
           cur_char = reader->getc();
         } else {
-          write_int_with_length(0, to_conv);
-          return READ_OK;
+          return MATCHING_FAILURE;
```
Comment (nickdesaulniers) [inline libc/src/stdio/scanf_core/int_converter.cpp:144]: Consider making this an early return if you reverse the conditional.

**Review by nickdesaulniers — APPROVED** [review]

File: libc/src/stdio/scanf_core/int_converter.cpp (hunk 2 above)
Comment (michaelrj-google) [reply libc/src/stdio/scanf_core/int_converter.cpp:144]: The `if (max_width > 1)` condition is a common pattern in scanf, so I'd rather keep the consistency and not reverse this conditional.

**Review by lntue — APPROVED** [review]

**Pull request merged**